raster_analysis/layer/window.py

```diff
@@ -0,0 +1,309 @@
+import numpy as np
+from rasterio import Affine
+from shapely.geometry import Polygon
+from numpy import ndarray
+from math import floor
+
+from raster_analysis.grid import get_raster_uri
+from raster_analysis.io import read_window
+
+from datetime import date, timedelta, datetime
+from typing import Union, Tuple
+
+Numeric = Union[int, float]
+WINDOW_SIZE = 4000
+
+
+class Window:
+    def __init__(self, layer: str, tile: Polygon):
+        self.layer: LayerInfo = LayerInfo(layer)
+        self.tile: Polygon = tile
+
+        self.data: ndarray = None
+        self.shifted_affine: Affine = None
+        self.no_data_value: Numeric = 0
+
+        self.data, self.shifted_affine, self.no_data_value = self.read(tile)
+        self._result: ndarray = None
+
+    def read(self, tile: Polygon) -> Tuple[np.ndarray, Affine, Numeric]:
+        data, shifted_affine, no_data_value = read_window(self.get_raster_uri(), tile)
+        print(f"MAX for {tile.bounds}: {data.max()}")
+
+        if data.size == 0:
+            data = np.zeros((WINDOW_SIZE, WINDOW_SIZE), dtype=np.uint8)
+
+        return data, shifted_affine, no_data_value
+
+    def sum(self, mean_area, mask, linear_index=None, index_counts=None):
+        data = np.extract(mask, self.data)
+
+        if linear_index is None and index_counts is None:
+            return data.sum()
+
+        return np.bincount(linear_index, weights=data, minlength=index_counts.size)
+
+    def clear(self):
+        self.data = []
+
+    @property
+    def result(self):
+        return self.layer.name, self._result.tolist()
+
+    @result.setter
+    def result(self, value):
+        self._result = value
+
+    def get_raster_uri(self):
+        return get_raster_uri(self.layer.layer, self.layer.data_type, self.tile)
+
+
+def get_window(
+    layer: str, tile: Polygon, start_date: datetime, end_date: datetime
+) -> Window:
+    if layer == "area__ha":
+        return AreaWindow(layer, tile)
+    elif layer == "alert__count":
+        return CountWindow(layer, tile)
+    elif layer == "tsc_tree_cover_loss_drivers__type":
+        return LossDriversWindow(layer, tile)
+    elif layer.startswith("umd_glad_alerts"):
+        return GladAlertsWindow(layer, tile, start_date, end_date)
+    elif layer.endswith("__year"):
+        return YearWindow(layer, tile, start_date, end_date)
+    elif layer.startswith("umd_tree_cover_density"):
+        return TcdWindow(layer, tile)
+    elif "__ha-1" in layer:
+        return AreaDensityWindow(layer, tile)
+    elif "whrc_aboveground_co2_emissions" in layer:
+        return CarbonEmissionsWindow(layer, tile)
+    else:
+        return Window(layer, tile)
```

Comment:
Only layers that need to be treated "differently"; e.g. glad alerts, which has a weird encoding, or layers like biomass density or emissions which need to be multiplied by a constant or a calculated from a different layer. 

The "__type" layers can be generalized to one subclass that looks up the appropriate mapping (or in the future, just gets it from the data API). 

It would be good though if in the future, all these "special" layers can have their rules defined somewhere in the data API (e.g. the CO2 emissions somehow being registered in the data API to say it's the biomass density layer multiplied by the area and a constant).